sdk/metrics/src/jmh/java/io/opentelemetry/sdk/metrics/MetricsBenchmarks.java

```diff
@@ -51,7 +51,7 @@ public class MetricsBenchmarks {
     }
   }
 
-  @State(Scope.Thread)
+  @State(Scope.Benchmark)
```

Comment:
Can you run this benchmark before and after so we can understand the impact? I understand its not apples to apples but it'd be good to understand how much of an impact fixing the concurrency has on the benchmarks.

Reply:
I did. I simply forgot to post the results.

I ran only for one method which I think is most interesting - eight threads one.

## Thread scope
```
Benchmark                                               (opBuilder)             (sdk)  Mode  Cnt   Score   Error  Units
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd          API_ONLY  avgt   10   1.068 ± 0.064  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd  SDK_NO_EXEMPLARS  avgt   10  13.721 ± 0.423  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd               SDK  avgt   10  31.719 ± 2.491  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd          API_ONLY  avgt   10   1.035 ± 0.030  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd  SDK_NO_EXEMPLARS  avgt   10  14.306 ± 0.590  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd               SDK  avgt   10  33.085 ± 2.970  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord          API_ONLY  avgt   10   4.530 ± 0.083  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord  SDK_NO_EXEMPLARS  avgt   10  46.169 ± 1.604  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord               SDK  avgt   10  86.183 ± 3.495  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord          API_ONLY  avgt   10   3.303 ± 0.020  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord  SDK_NO_EXEMPLARS  avgt   10  40.490 ± 1.017  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord               SDK  avgt   10  82.717 ± 5.601  ns/op
```

## Benchmark scope
```
Benchmark                                               (opBuilder)             (sdk)  Mode  Cnt    Score    Error  Units
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd          API_ONLY  avgt   10    1.013 ±  0.011  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd  SDK_NO_EXEMPLARS  avgt   10   15.819 ±  1.712  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet         LongCounterAdd               SDK  avgt   10   24.989 ±  1.531  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd          API_ONLY  avgt   10    1.024 ±  0.022  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd  SDK_NO_EXEMPLARS  avgt   10   16.228 ±  0.041  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet       DoubleCounterAdd               SDK  avgt   10   21.785 ±  0.305  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord          API_ONLY  avgt   10    4.423 ±  0.021  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord  SDK_NO_EXEMPLARS  avgt   10  522.427 ±  7.888  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet  DoubleHistogramRecord               SDK  avgt   10  576.247 ± 12.840  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord          API_ONLY  avgt   10    3.657 ±  0.087  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord  SDK_NO_EXEMPLARS  avgt   10  521.295 ±  4.485  ns/op
MetricsBenchmarks.eightThreadsCommonLabelSet    LongHistogramRecord               SDK  avgt   10  564.783 ± 12.550  ns/op
```

For LongCounter SDK, thread scope is 31.1 ns/operation while benchmark scope is 24 ns/operation - sharing the same storage (SDK) made the operation be faster. 
On the SDK histogram, making it shared structure made it x6 slower from ~80 ns/op to ~560 ns/op.

I think it simply revealed we need to examine the test and cpu profile (flamegraph) and see the bottlenecks. 

If you compare each: In thread-scope, long counter SDK is 31 ns/op vs long histogram SDK which is 82, so histogram is x2.5 slower, but when you are in benchmark scope, actually sharing the storage, you see that it's ~25 ns/op vs 564 ns/op, making histogram x22 slower!

Reply:
> For LongCounter SDK, thread scope is 31.1 ns/operation while benchmark scope is 24 ns/operation 

I bet the slower time when it was thread scoped was due to the overhead of having more allocations from having several copies of the SDK running.

> I think it simply revealed we need to examine the test and cpu profile (flamegraph) and see the bottlenecks.

👍 Thanks for running it.